swift/ql/src/queries/Security/CWE-328/WeakPasswordHashing.qhelp

```diff
@@ -0,0 +1,96 @@
+<!DOCTYPE qhelp PUBLIC
+        "-//Semmle//qhelp//EN"
+        "qhelp.dtd">
+<qhelp>
+    <overview>
+        <p>
+            Using a insufficiently computationally hard hash function can leave data
+            vulnerable, and should not be used for password hashing.
+        </p>
+
+        <p>
+            A strong cryptographic hash function should be resistant to:
+        </p>
+        <ul>
+            <li>
+                <strong>Pre-image attacks</strong>. If you know a hash value <code>h(x)</code>,
+                you should not be able to easily find the input <code>x</code>.
+            </li>
+            <li>
+                <strong>Collision attacks</strong>. If you know a hash value <code>h(x)</code>,
+                you should not be able to easily find a different input
+                <code>y</code>
+                with the same hash value <code>h(x) = h(y)</code>.
+            </li>
+            <li>
+                <strong>Brute force</strong>. If you know a hash value <code>h(x)</code>,
+                you should not be able to find an input <code>y</code> that computes to that hash value
+                using brute force attacks without significant computational effort.
+            </li>
+        </ul>
+
+        <p>
+            All of MD5, SHA-1, SHA-2 and SHA-3 are weak against offline brute forcing, since they are not computationally hard. This includes SHA-224, SHA-256, SHA-384 and SHA-512, which are in the SHA-2 family.
+        </p>
+
+        <p>
+            Password hashing algorithms are designed to be slow and/or memory intenstive to compute, which makes brute force attacks more difficult.
+        </p>
+
+    </overview>
+    <recommendation>
+
+        <p>
+            Ensure that for password storage you should use a computationally hard cryptographic hash function, such as:
+        </p>
+
+        <ul>
+            <li>
+                Argon2
+            </li>
+            <li>
+                scrypt
+            </li>
+            <li>
+                bcrypt
+            </li>
+            <li>
+                PBKDF2
+            </li>
+        </ul>
+
+    </recommendation>
+    <example>
+
+        <p>
+            The following examples show a function that hashes a password using a cryptographic hashing algorithm.
+
+            In the first case the SHA-512 hashing algorithm is used. It is vulnerable to offline brute force attacks:
+        </p>
```

Comment:
```suggestion
        <p>
            The following examples show two versions of the same function. In both cases, a password is hashed using a cryptographic hashing algorithm.

            In the first case, the SHA-512 hashing algorithm is used. It is vulnerable to offline brute force attacks:
        </p>
```

I don't think these edits are essential, but I thought this might help clarify the content a little.

You might want to mark this as "BAD" and the second as "GOOD" as is often the case with query examples, but up to you.

Reply:
I have added GOOD and BAD comments as well, for this query and the existing weak sensitive data hashing query (that is very similar to this one).